### PR TITLE
Document api

### PR DESCRIPTION
**Pull request by mnunberg**
Document API refactor; this fixes some leaks; the rest of the leak-fixing is in the plug-leaks branch.

### CODE REVIEW

File: src/document_basic.c [hunk 1]
```diff
@@ -1,94 +1,181 @@
...
+  doc->payloadSize = 0;
 }
 
 void Document_AddField(Document *d, const char *fieldname, RedisModuleString *fieldval,
                        uint32_t typemask) {
   d->fields = realloc(d->fields, (++d->numFields) * sizeof(*d->fields));
   DocumentField *f = d->fields + d->numFields - 1;
   f->indexAs = typemask;
-  f->name = strdup(fieldname);  // free_detached called on this later on..
+  f->name = fieldname;
```
Comment (MeirShpilraien) [inline src/document_basic.c]: Why not duplicate? If the api state const var then its better to duplicate and not take ownership.

File: src/document_basic.c [hunk 2]
```diff
@@ -1,94 +1,181 @@
...
+  if (!k || RedisModule_KeyType(k) != REDISMODULE_KEYTYPE_HASH) {
+    goto done;
+  }
+
+  size_t nitems = RedisModule_ValueLength(k);
+  if (nitems == 0) {
+    goto done;
+  }
+
+  Document_MakeStringsOwner(doc);
```
Comment (MeirShpilraien) [inline src/document_basic.c]: We should come with a better name to this function, maybe take ownership?

Comment (mnunberg) [reply src/document_basic.c]: Well, it doesn't actually *take* ownership of the strings; it makes it own its own copy of the strings

**Comment by mnunberg** [comment]
@MeirShpilraien addressed the comments; let's merge this so we can merge the bigger one

**Pull request merged**